### PR TITLE
fix(lambda): allow ArnPrincipal for lambda grantInvoke

### LINKED ISSUE #3264: Invalid principal type for Lambda permission statement: ArnPrincipal. Supported: AccountPrincipal, ServicePrincipal
I'm trying to allow a lambda function in another account to be able to invoke my CDK generated lambda function. This works through the CLI like so:

`aws lambda add-permission --function-name=myFunction --statement-id=ABoldStatement --action=lambda:InvokeFunction --principal=arn:aws:iam::<account_id>:role/a_lambda_execution_role`

But CDK doesn't seem to allow me to add an ArnPrincipal doing something like this:

`myFunction.grantInvoke(new iam.ArnPrincipal(props.myARN))`

With the error `Invalid principal type for Lambda permission statement: ArnPrincipal. Supported: AccountPrincipal, ServicePrincipal`

Seems to be around here https://github.com/awslabs/aws-cdk/blob/4a0272db5eb9aae5f440bb0813fdbad6514b60c3/packages/%40aws-cdk/aws-lambda/lib/function-base.ts#L278

I can see about adding this if we think it's correct to add.

### PR DESCRIPTION
**Pull request by IainCole**
Fixes #3264 

I'm trying to allow a lambda function in another account to be able to invoke my CDK generated lambda function. This works through the CLI like so:

    aws lambda add-permission --function-name=myFunction --statement-id=ABoldStatement --action=lambda:InvokeFunction --principal=arn:aws:iam::{account_id}:role/a_lambda_execution_role

But CDK doesn't seem to allow me to add an ArnPrincipal doing something like this:

    myFunction.grantInvoke(new iam.ArnPrincipal(props.myARN))

With the error:

    Invalid principal type for Lambda permission statement: ArnPrincipal. Supported: AccountPrincipal, ServicePrincipal

This PR allows ArnPrincipal to be passed to lambda.grantInvoke.

There might be some additional validation required on the exact ARN as I believe only some ARNs are supported by lambda add-permission

### CODE REVIEW

**Comment by IainCole** [comment]
@NGL321 PR as requested

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts [hunk 1]
```diff
@@ -285,6 +285,10 @@ export abstract class FunctionBase extends Resource implements IFunction {
       return (principal as iam.ServicePrincipal).service;
     }
 
+    if (`arn` in principal) {
+      return (principal as iam.ArnPrincipal).arn;
+    }
+
     throw new Error(`Invalid principal type for Lambda permission statement: ${principal.constructor.name}. ` +
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda/lib/function-base.ts]: Update this to include `ArnPrincipal`

File: packages/@aws-cdk/aws-lambda/test/test.lambda.ts [hunk 2]
```diff
@@ -186,19 +186,6 @@ export = {
       test.done();
     },
 
-    'fails if the principal is not a service or account principals'(test: Test) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda/test/test.lambda.ts]: Leave this test, but use a different type of principal (even mock one that doesn't have any of the options)

File: packages/@aws-cdk/aws-lambda/lib/function-base.ts (hunk 1 above)
Comment (IainCole) [reply packages/@aws-cdk/aws-lambda/lib/function-base.ts]: 🤦‍♂

File: packages/@aws-cdk/aws-lambda/test/test.lambda.ts (hunk 2 above)
Comment (IainCole) [reply packages/@aws-cdk/aws-lambda/test/test.lambda.ts]: I used OrganizationPrincipal

**Pull request merged**